### ISSUE TITLE
Don't import all modules from torchmetrics root

### ISSUE THREAD

**Issue by ananthsub** [PyTorchLightning/metrics#459]
## 🚀 Feature

@SkafteNicki could we not import all metrics at the root module? https://github.com/PyTorchLightning/metrics/blob/master/torchmetrics/__init__.py

### Motivation

Reasons against this:
- This slows down import times
- Not all requirements might be installed. For instance, [torch-fidelity is required for FID under torchmetrics.image](https://github.com/PyTorchLightning/metrics/blob/00f0256db37406df09c18ddc18f4eca99ef7d59e/torchmetrics/image/fid.py#L26-L27). But torchmetrics itself doesn't require these dependencies to be installed. 

### Pitch
Keep metric implementation imports at the submodule level, and ensure the root torchmetrics is used for only the essential classes (e.g. `Metric`)

**Comment by Borda** [comment]
> * This slows down import times

do you have any benchmark, how much slower it is?

> * Not all requirements might be installed.

this can be easily solved by conditional import... :rabbit:

**Comment by SkafteNicki** [comment]
Small benchmark executing the statement "import torchmetrics as tm" 10 times
On master: 1.8266461610794067s+-0.033780458858975175 s
PR #463: 1.5197294712066651s+-0.07895099137983734 s

**Comment by ananthsub** [comment]
@Borda it's even more pronounced when using python with tons of shared libraries through a PAR or XAR file

**Comment by awaelchli** [comment]
For comparison: sklearn has also all metrics importat at the top level at sklearn.metrics. They also rely on numpy like we do on pytorch. 
On the other hand, as the inventory of metrics grow, it is possible that there can be collisions in different domains where metrics have the same name but are defined very differently.

**Comment by carmocca** [comment]
I think this is worth it considering the vast majority of users import 1-3 metrics at most. Loading the entire package for this is overkill

**Comment by tchaton** [comment]
I would align with Ananth there. 

As the project grows, I believe there is need to provide better organizations and namespace, especially it saves loading time.

A benefit from this approach might be to incentivise users to check out other namespace and discover new metrics.

Best,
T.C

**Comment by Borda** [comment]
> As the project grows, I believe there is need to provide better organizations and namespace, especially it saves loading time.

I don't see any speed degradation, @ananthsub can you provide some benchmark?
@tchaton this is breaking chnage with any back compatibility for all users so this shall be well justified =)

**Comment by akihironitta** [comment]
> still don't see why top level imports could be problem if there is only torch dependency....

As @Borda mentioned, if there're no optional packages installed, loading time doesn't change, but if there are all of them (in a rare situation maybe?), it's around 4x longer than #463 if I ran a benchmark correctly in the notebook below.

- With no optional dependencies
  - master: 916 ms
  - [#463](https://github.com/PyTorchLightning/metrics/pull/463): 912 ms
- With all optional dependencies
  - master: 4.02 s
  - [#463](https://github.com/PyTorchLightning/metrics/pull/463): 914 ms

*best of 5 runs of `import torchmetrics` each*

Notebook: https://colab.research.google.com/drive/1PY2Nmba8fpxHpyAPAwLpaCdbRrO-kejj?usp=sharing

**Comment by ananthsub** [comment]
@Borda  here are some sample timings in our developer environment running this import script:
```
#!/usr/bin/env python3

if __name__ == "__main__":
    import torchmetrics
```

in our development environment:

Main branch:
```
real    0m11.831s
user    0m10.876s
sys     0m0.939s
```

with #463:
```
real    0m1.035s
user    0m0.832s
sys     0m0.175s
```

As the main issue states, we'd like to avoid importing from modules given the optional requirements. If someone is only defining a custom Metric, they shouldn't end up importing the whole package. Otherwise, the import time will shoot up as more dependencies are added. 

I completely understand that backwards compatibility is critical. I know that updating import paths across lots of projects is not fun, having had to personally do this in the past. At the same time, torchmetrics is pre-1.0 release, so this is the time to make breaking changes while the API is being solidifed and import patterns are being decided. If you have suggestions on how we could not by default import all modules while preserving backward compatibility, I'd be happy to learn of alternatives!

**Comment by Borda** [comment]
@ananthsub so how about if we keep in the package root Metric which is truly dependent only on PT and all the needs other/external dependencies will be imported from its sub-packages? Pls understand that otherwise, this is a significant break change with any back compatibility and any warning...

**Comment by ananthsub** [comment]
> @ananthsub so how about if we keep in the package root Metric which is truly dependent only on PT and all the needs other/external dependencies will be imported from its sub-packages? Pls understand that otherwise, this is a significant break change with any back compatibility and any warning...

do you think this pattern is obvious to users for why some metrics are imported from submodules while others are imported at the root?

**Comment by Borda** [comment]
> do you think this pattern is obvious to users for why some metrics are imported from submodules while others are imported at the root?

no, it isn't, but also doe not harm actual users :rabbit: 
to be systematic we would keep in root the basic metrics like classification or regression and move out domain-specific like CV or NLP

**Comment by SkafteNicki** [comment]
Last week I discussed with @Borda about having some environment variable that can be set to indicate if the user want the faster import pattern:
```python
if os.environ.get('TM_FAST'):
    #import as in PR 463
else:
    #import as we do it now
```
it is not a pretty solution, but it is a compromise.
